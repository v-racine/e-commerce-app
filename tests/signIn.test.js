const { Config } = require('../src/config/config');

Config.Get(process.env);

const AppFactory = require('../src/app');
const request = require('supertest');

describe('sign in', () => {
  const mockUsersRepo = {};

  const app = AppFactory({
    usersRepo: mockUsersRepo,
  });

  describe('when: the user passes in something that is not an email for email', () => {
    let rsp;

    beforeEach(async () => {
      rsp = await request(app).post('/signin').send({ email: 'asdf', password: 'testing' });
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test("then: we return 'Must be a valid email' html", async () => {
      const status = rsp.status;
      expect(status).toBe(200);

      const text = rsp.text;
      expect(text).toMatchSnapshot();
    });
  });

  describe('when: the db does not already contain that email', () => {
    let rsp;

    beforeEach(async () => {
      mockUsersRepo.getOneBy = jest.fn().mockImplementation(() => {
        return new Promise((resolve) => {
          resolve(undefined);
        });
      });

      rsp = await request(app).post('/signin').send({ email: 'testing1@test.com' });
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test("then: we return a 'Email not found'", async () => {
      const status = rsp.status;
      expect(status).toBe(200);

      const text = rsp.text;
      expect(text).toBe('Email not found');

      expect(mockUsersRepo.getOneBy).toHaveBeenCalledWith({ email: 'testing1@test.com' });
    });
  });

  describe('when: the passwords do not match', () => {
    let rsp;

    beforeEach(async () => {
      mockUsersRepo.getOneBy = jest.fn().mockImplementation(() => {
        return new Promise((resolve) => {
          resolve({
            email: 'testing2@test.com',
            password:
              'c39c38f8856779b2271931a7bf0b9b9002f452a8450b81e54e12a77e9025653e401ec28c4839bcfba384d90a94f13caf4fa8f52104cf72fd04f0dec2272c1330.9a4dfb72bbf1f3ef',
          });
        });
      });

      rsp = await request(app)
        .post('/signin')
        .send({ email: 'testing2@test.com', password: 'testing2' });
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('then: we return "Invalid password"', async () => {
      const status = rsp.status;
      expect(status).toBe(200);

      const text = rsp.text;
      expect(text).toBe('Invalid password');

      expect(mockUsersRepo.getOneBy).toHaveBeenCalledWith({ email: 'testing2@test.com' });
    });
  });

  describe('when: the db contains the email and the passwords match', () => {
    let rsp;

    beforeEach(async () => {
      mockUsersRepo.getOneBy = jest.fn().mockImplementation(() => {
        return new Promise((resolve) => {
          resolve({
            email: 'testing5@test.com',
            password:
              'c39c38f8856779b2271931a7bf0b9b9002f452a8450b81e54e12a77e9025653e401ec28c4839bcfba384d90a94f13caf4fa8f52104cf72fd04f0dec2272c1330.9a4dfb72bbf1f3ef',
          });
        });
      });

      rsp = await request(app)
        .post('/signin')
        .send({ email: 'testing5@test.com', password: 'papillon' });
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('then: we redirect to products list', async () => {
      const status = rsp.status;
      expect(status).toBe(302);

      const text = rsp.text;
      expect(text).toBe('Found. Redirecting to /admin/products');

      expect(rsp.get('Set-Cookie').length).not.toBe(0);

      expect(mockUsersRepo.getOneBy).toHaveBeenCalledWith({ email: 'testing5@test.com' });
    });
  });
});
