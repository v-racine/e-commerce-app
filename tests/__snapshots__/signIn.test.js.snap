// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sign in when: the user passes in something that is not an email for email then: we return 'Must be a valid email' html 1`] = `
"
    <!DOCTYPE html>
      <html lang="en">
      <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Shop</title>
        <link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" rel="stylesheet">
        <link href="/css/main.css" rel="stylesheet">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bulma/0.7.5/css/bulma.min.css"></link>
      </head>

      <body class="admin">
        <header>
          <nav class="navbar navbar-bottom">
            <div class="container navbar-container">
              <div>
                <a href="/admin/products">
                  <h3 class="title">Admin Panel</h3>
                </a>
              </div>
              <div class="navbar-item">
                <div class="navbar-buttons">
                  <div class="navbar-item">
                    <a href="/admin/products"><i class="fa fa-star"></i> Products</a>
                  </div>
                </div>
              </div>
            </div>
          </nav>
        </header>
        <div class="container">
          
      <div class="container">
        <div class="columns is-centered">
          <div class="column is-one-quarter">
            <form method="POST">
              <h1 class="title">Sign in</h1>
              <div class="field">
                <label class="label">Email</label>
                <input required class="input" placeholder="Email" name="email" />
                <p class="help is-danger">Must be a valid email</p>
              </div>
              <div class="field">
                <label class="label">Password</label>
                <input required class="input" placeholder="Password" name="password" type="password" />
                <p class="help is-danger"></p>
              </div>
              <button class="button is-primary">Submit</button>
            </form>
            <a href="/signup">Need an account? Sign Up</a>
          </div>
        </div>
      </div>
    
        </div>
      </body>
    </html>
  "
`;
